internal/rootcoord/root_coord.go

```diff
@@ -2428,30 +2458,40 @@ func (c *Core) checkCompleteIndexLoop(ctx context.Context, ti *datapb.ImportTask
 	for {
 		select {
 		case <-c.ctx.Done():
-			log.Info("(in loop)context done, exiting checkCompleteIndexLoop",
+			log.Info("(in check complete index loop) context done, exiting checkCompleteIndexLoop",
 				zap.Int64("task ID", ti.GetId()))
 			return
 		case <-ticker.C:
-			log.Info("(in loop)check segments' index states", zap.Int64("task ID", ti.GetId()))
+			log.Info("(in check complete index loop) check segments' index states", zap.Int64("task ID", ti.GetId()))
 			if ct, err := c.CountCompleteIndex(ctx, colName, ti.GetCollectionId(), segIDs); err == nil &&
 				segmentsOnlineReady(ct, len(segIDs)) {
-				log.Info("(in loop)segment indices are ready",
+				log.Info("segment indices are ready",
 					zap.Int64("task ID", ti.GetId()),
 					zap.Int("total # of segments", len(segIDs)),
 					zap.Int("# of segments with index ready", ct))
-				c.importManager.bringSegmentsOnline(ti)
+				c.bringSegmentsOnline(ctx, segIDs)
 				return
 			}
 		case <-expireTicker.C:
-			log.Info("(in loop)waited for sufficiently long time, bring segments online",
+			log.Info("(in check complete index loop) waited for sufficiently long time, bring segments online",
 				zap.Int64("task ID", ti.GetId()))
-			c.importManager.bringSegmentsOnline(ti)
+			c.bringSegmentsOnline(ctx, segIDs)
 			return
 		}
 	}
 }
 
-// segmentsOnlineReady returns true if segments are ready to go up online (a.k.a. searchable).
+// bringSegmentsOnline brings the segments online so that data in these segments become searchable
+// it is done by changing segments' states from `importing` to `flushed`.
+func (c *Core) bringSegmentsOnline(ctx context.Context, segIDs []UniqueID) {
+	log.Info("bringing import task's segments online!", zap.Any("segment IDs", segIDs))
+	for _, id := range segIDs {
+		// Explicitly mark segment states `flushed`.
+		c.CallUpdateSegmentStateService(ctx, id, commonpb.SegmentState_Flushed)
+	}
+}
```

Comment:
Looks like this operation is not atomic.

Reply:
Yes it should be atomic. I left a TODO here and will address it in the next PR.